src/Symfony/Component/Utf8/Utf8Trait.php

```diff
@@ -0,0 +1,397 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Utf8;
+
+use Symfony\Component\Utf8\Exception\InvalidArgumentException;
+
+/**
+ * @internal
+ */
+trait Utf8Trait
+{
+    private static $commonCaseFold = array(
+        array('µ','ſ',"\xCD\x85",'ς',"\xCF\x90","\xCF\x91","\xCF\x95","\xCF\x96","\xCF\xB0","\xCF\xB1","\xCF\xB5","\xE1\xBA\x9B","\xE1\xBE\xBE"),
+        array('μ','s','ι',       'σ','β',       'θ',       'φ',       'π',       'κ',       'ρ',       'ε',       "\xE1\xB9\xA1",'ι'),
+    );
+
+    private $string = '';
+
+    public static function fromString(string $string)
+    {
+        if (!preg_match('//u', $string)) {
+            throw new InvalidArgumentException('Given string is not a valid UTF-8 encoded string.');
+        }
+
+        $instance = new static();
+        $instance->string = $string;
+
+        return $instance;
+    }
+
+    public static function fromCodePoint(int ...$codes)
+    {
+        $string = '';
+        foreach ($codes as $code) {
+            if (0x80 > $code %= 0x200000) {
+                $string .= chr($code);
+            } elseif (0x800 > $code) {
+                $string .= chr(0xC0 | $code >> 6).chr(0x80 | $code & 0x3F);
+            } elseif (0x10000 > $code) {
+                $string .= chr(0xE0 | $code >> 12).chr(0x80 | $code >> 6 & 0x3F).chr(0x80 | $code & 0x3F);
+            } else {
+                $string .= chr(0xF0 | $code >> 18).chr(0x80 | $code >> 12 & 0x3F).chr(0x80 | $code >> 6 & 0x3F).chr(0x80 | $code & 0x3F);
+            }
+        }
+
+        $instance = new static();
+        $instance->string = $string;
+
+        return $instance;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function explode(string $delimiter, int $limit = null): array
+    {
+        if ('' === $delimiter) {
+            throw new InvalidArgumentException('Passing an empty delimiter is not supported by this method. Use getIterator() method instead.');
+        }
+
+        $chunks = array();
+        foreach (preg_split('/'.preg_quote($delimiter).'/u', $this->string, $limit ?: -1) as $i => $string) {
+            $chunks[$i] = $chunk = clone $this;
+            $chunk->string = $string;
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtolower($this->string, 'UTF-8');
```

Comment:
This will not yield the desired result in some languages. Classical example for this is Turkish:

```php
assert(mb_strtolower('DİNÇ') === 'dinç');
assert(mb_strtolower('DINÇ') === 'dınç');
```

The latter will not be converted correctly.

```php
assert(transliterator_transliterate('tr-Lower', 'DİNÇ') === 'dinç');
assert(transliterator_transliterate('tr-Lower', 'DINÇ') === 'dınç');
```

This works as expected. Of course the same is true for `toUpperCase`.